Add visualization module for quantum process

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
This PR adds visualization tool for quantum process.

close #3374 

### Details and comments
```python
from qiskit import quantum_info as qi

qc = qiskit.QuantumCircuit(2)
qc.cx(0, 1)

plot_process_city(qi.Chi(qc).data, alpha=0.7, figsize=(15, 6))
```
This code returns below matplot figure object.
![index](https://user-images.githubusercontent.com/39517270/68018633-f2eaf580-fcdc-11e9-9f8e-0dece5095c14.png)

This new function reuses `plot_state_city` but different axis label. To reduce maintenance cost, original core functionality of visualization is moved to private function `_plot_matrix_city` and this is called both from `plot_state_city` and `plot_process_city`.

chriseclectic commented on qiskit/visualization/state_visualization.py:
```diff
@@ -330,19 +328,140 @@ def plot_state_city(rho, title="", figsize=None, color=None,
+    part of the channel.
+
+    Args:
+        channel (ndarray): Numpy array representing quantum channel.
```
Can you make this `channel (QuantumChannel)` so it takes channel objects, and converts them to Chi matrix

chriseclectic commented on qiskit/visualization/state_visualization.py:
```diff
@@ -330,19 +328,140 @@ def plot_state_city(rho, title="", figsize=None, color=None,
+    # get the labels
+    row_names = list(map(''.join, itertools.product(basis, repeat=num)))
+    column_names = list(map(''.join, itertools.product(basis, repeat=num)))
+
```
To convert other channel representations into Chi matrix you can add here

```python
mat = Chi(channel).data
fig = _plot_matrix_city(mat=mat, ...)
```
```

chriseclectic commented on qiskit/visualization/state_visualization.py:
```diff
@@ -330,19 +328,140 @@ def plot_state_city(rho, title="", figsize=None, color=None,
+        return fig
+
+
+def plot_process_city(channel, title="", figsize=None, color=None,
```
I would rename this to `plot_quantum_channel_city` and make it work for any representation of the `QuantumChannel` classes by converting to the Chi matrix

chriseclectic commented on qiskit/visualization/state_visualization.py:
```diff
@@ -330,19 +328,140 @@ def plot_state_city(rho, title="", figsize=None, color=None,
+        return fig
+
+
+def _plot_matrix_city(mat, row_names, column_names, oper_name,
```
This could be a useful function to make available to users with signature:

```python
def plot_matrix_city(mat, row_labels=None, col_labels=None, oper_name='matrix',  ...)
    """Plot the cityscape of arbitrary matrix.
    Plot two 3d bar graphs (two dimensional) of the real and imaginary
    part of a complex matrix.
    Args:
        mat (ndarray): Numpy array of matrix to visualize.
        row_labels (list(str)): Labels for row. If `None` use integer labels.
        column_names (list(str)): Labels for column. If `None` use integer labels.
        row_labels (list(str)): Labels for row. If `None` use integer labels.
        column_names (list(str)): Labels for column. If `None` use integer labels.
        <rest the same>  
```

Then `plot_state_city` and the new function will just be this with appropriate axis labels

chriseclectic commented on qiskit/visualization/state_visualization.py:
```diff
@@ -330,19 +328,140 @@ def plot_state_city(rho, title="", figsize=None, color=None,
+def plot_process_city(channel, title="", figsize=None, color=None,
+                      alpha=1, ax_real=None, ax_imag=None,
+                      oper_name=None, basis=None):
+    """Plot the cityscape of quantum channel.
```
City plot of the Chi-matrix representation of a quantum channel.

nkanazawa1989 commented:
@chriseclectic thanks for reviewing this! I updated interface to receive `QuantumChannel` object instead of ndarray. Also added `plot_quantum_channel_hinton` function to visualize Chi-matrix as a Hinton diagram in the same manner, by a request from private discussion.

Actually replacing both _state_ and _quantum_channel_ interfaces by single public _matrix_ interface is good direction to make the code simpler, but deprecating existing public module is not easy. This is why I keep _state_ interface in an original fashion.

taalexander commented:
Could you please add a reno @nkanazawa1989?

nonhermitian commented:
Why can't the input simply be the channel object, rather than the user having to call `.data` on the channel?

nkanazawa1989 commented:
@taalexander it's already in here [releasenotes/notes/add-process-matrix-plot-12f9456fcd47be52.yaml](https://github.com/Qiskit/qiskit-terra/blob/45acb2d86de338ffb84aa52064e80662ad4d3a33/releasenotes/notes/add-process-matrix-plot-12f9456fcd47be52.yaml)

@nonhermitian pr was updated to plot quantum object directly. given object is internally converted into Choi representation.

```
import qiskit

qc = qiskit.QuantumCircuit(2)
qc.cx(0, 1)

qiskit.visualization.state_visualization.plot_quantum_channel_hinton(qc)
```
![image](https://user-images.githubusercontent.com/39517270/70305155-0c84dd00-1847-11ea-844f-d367f45232f5.png)

nonhermitian commented on releasenotes/notes/add-process-matrix-plot-12f9456fcd47be52.yaml:
```diff
@@ -0,0 +1,28 @@
+    This function visualize Chi-matrix given by arbitrary quantum channel object
+    as a Hinton diagram, for example::
+
+        from qiskit import QuantumCircuit
```
These examples should be in the docstrings as well.

nkanazawa1989 replied:
it's already in here
https://github.com/Qiskit/qiskit-terra/blob/45acb2d86de338ffb84aa52064e80662ad4d3a33/qiskit/visualization/state_visualization.py#L152-L163

nonhermitian commented on qiskit/visualization/state_visualization.py:
```diff
@@ -330,19 +434,138 @@ def plot_state_city(rho, title="", figsize=None, color=None,
+    part of the quantum channel.
+
+    Args:
+        channel (QuantumChannel): Arbitrary QuantumChannel object to visualize.
```
This should say it takes a QuantumCircuit or channel since that it now what it does.

nonhermitian commented:
Should the Honton diagram have rectangles like it shows above?  Perhaps use an edge color or reduce the square size so that the matrix entries do not merge like that?  Also, the docs need to be in the new type-hints format.

nkanazawa1989 commented:
@nonhermitian thanks for the good suggestion. I updated docstring, typing and edge color.

now the plot like this
![image](https://user-images.githubusercontent.com/39517270/75623322-2a66a180-5bec-11ea-99b1-b5f932e822f1.png)

In addition, I added `auto_scale` option (which is forced before). when comparing two quantum channels, eg ideal vs expt result, this forced auto scale is very annoying. by specifying `auto_scale=False`,
![image](https://user-images.githubusercontent.com/39517270/75623352-88938480-5bec-11ea-8afa-aa4169aabb62.png)
usually reconstructed channel includes tiny z* components and difference of plot range sometime makes it difficult to compare it with ideal channel.

unfortunately the linter causes many cyclic import error and I've not found the reason of it.

nonhermitian commented:
Perhaps a colorbar and it should be good to go.

nkanazawa1989 commented:
the diagram has only two colors, ie 'white' for positive val, 'black' for negative val, and magnitude of component is visualized by size of squares. I'll put this info into the docstring instead of color bar (since there is no intermediate colors).

nkanazawa1989 commented:
I also fixed cyclic-import error. Maybe it is good tips how to fix.

If you want to import `QuantumCircuit` to your module for type hinting, you should
```
from qiskit.circuit import QuantumCircuit
```

please note that
```
from qiskit import QuantumCircuit
```
causes cyclic import error in many cases.

nkanazawa1989 commented:
@nonhermitian the PR passed all checks and it can be merged now. please check new docstrings if they are what you intended.

ajavadia commented:
@chriseclectic @nonhermitian ping

ajavadia commented on qiskit/visualization/state_visualization.py:
```diff
@@ -97,19 +127,178 @@ def plot_state_hinton(rho, title='', figsize=None, ax_real=None, ax_imag=None):
+
+
+def plot_quantum_channel_hinton(
+        channel: Union[QuantumChannel, QuantumCircuit],
```
I like accepting QuantumChannel directly but a lot of other plotting methods take ndarray. So I think we should support both for consistency. The ones being updated here (https://github.com/Qiskit/qiskit-terra/pull/4324/files) also preserve the ndarray functionality still. Maybe we deprecate this altogether but while they exist I think they should exist on all drawers.

nkanazawa1989 replied:
Thanks for the feedback. Actually the function is still able to receive ndarray, but we cannot guarantee the input is Chi matrix representation.

For example,
```
zx = quantum_info.Operator.from_label('ZX')
zx_array = zx.data

plot_quantum_channel_hinton(zx_array)
```
will output meaningless plot (not like a quantum state, there are multiple channel representations in Qiskit).

So I explicitly added ndarray to the type hint and also added a warning when the array is given.

taalexander commented:
This is coming up on a year of being open. Should this be closed?

taalexander commented:
Closing as stale. Feel free to reopen.
